src/librbd/cache/ReplicatedWriteLog.h

```diff
@@ -0,0 +1,332 @@
+  ReplicatedWriteLog(ImageCtxT &image_ctx, ImageCacheState<ImageCtxT>* cache_state);
+  ~ReplicatedWriteLog();
+  ReplicatedWriteLog(const ReplicatedWriteLog&) = delete;
+  ReplicatedWriteLog &operator=(const ReplicatedWriteLog&) = delete;
+
+  /// client AIO methods
+  void aio_read(Extents&& image_extents, ceph::bufferlist *bl,
+		int fadvise_flags, Context *on_finish) override;
+  void aio_write(Extents&& image_extents, ceph::bufferlist&& bl,
+		 int fadvise_flags, Context *on_finish) override;
+  void aio_discard(uint64_t offset, uint64_t length,
+		   uint32_t discard_granularity_bytes,
+		   Context *on_finish) override;
+  void aio_flush(Context *on_finish) override;
+  void aio_writesame(uint64_t offset, uint64_t length,
+		     ceph::bufferlist&& bl,
+		     int fadvise_flags, Context *on_finish) override;
+  void aio_compare_and_write(Extents&& image_extents,
+			     ceph::bufferlist&& cmp_bl, ceph::bufferlist&& bl,
+			     uint64_t *mismatch_offset,int fadvise_flags,
+			     Context *on_finish) override;
+
+  /// internal state methods
+  void init(Context *on_finish) override;
+  void shut_down(Context *on_finish) override;
+  void invalidate(Context *on_finish);
+  void flush(Context *on_finish) override;
+
+private:
+  std::atomic<bool> m_initialized = {false};
+  ReplicatedWriteLogInternal *m_internal;
+  const char* rwl_pool_layout_name;
+
+  ImageCtxT &m_image_ctx;
+
+  std::string m_log_pool_name;
+  bool m_log_is_poolset = false;
+  uint64_t m_log_pool_config_size; /* Configured size of RWL */
+  uint64_t m_log_pool_actual_size = 0; /* Actual size of RWL pool */
+
+  uint32_t m_total_log_entries = 0;
+  uint32_t m_free_log_entries = 0;
+
+  std::atomic<uint64_t> m_bytes_allocated = {0}; /* Total bytes allocated in write buffers */
+  uint64_t m_bytes_cached = 0;    /* Total bytes used in write buffers */
+  uint64_t m_bytes_dirty = 0;     /* Total bytes yet to flush to RBD */
+  uint64_t m_bytes_allocated_cap = 0;
+
+  ImageWriteback<ImageCtxT> m_image_writeback;
+
+  /*
+   * When m_first_free_entry == m_first_valid_entry, the log is
+   * empty. There is always at least one free entry, which can't be
+   * used.
+   */
+  uint64_t m_first_free_entry = 0;  /* Entries from here to m_first_valid_entry-1 are free */
+  uint64_t m_first_valid_entry = 0; /* Entries from here to m_first_free_entry-1 are valid */
+
+  /* Starts at 0 for a new write log. Incremented on every flush. */
+  uint64_t m_current_sync_gen = 0;
+  /* Starts at 0 on each sync gen increase. Incremented before applied
+     to an operation */
+  uint64_t m_last_op_sequence_num = 0;
+  /* All writes bearing this and all prior sync gen numbers are flushed */
+  uint64_t m_flushed_sync_gen = 0;
+
+  /* Acquire locks in order declared here */
+
+  mutable ceph::mutex m_timer_lock; /* Used with and by m_timer */
+  /* Used for most synchronization */
+  mutable ceph::mutex m_lock;
+
+  Contexts m_flush_complete_contexts;
+  Finisher m_persist_finisher;
```

Comment:
The finisher stuff is a relic, and can all go. These didn't get the result we wanted, but the private thread pool did.

We do need the private thread pool. We can't hit our latency targets with shared threads. We can't control what's ahead of us in shared WQs, or other WQs sharing the thread pool. Basically the RWL WQs and threads are for gravel, and the librbd shared WQ has boulders in it.

During startup and shutdown we rely on these separate threads to avoid circular dependencies on things in the shared WQ that block. Other than that RWL works without the private threads, but performance is awful.

Long term I was hoping to open a discussion about WQ and thread pool features that would allow RWL to avoid private threads. That's hard to do until the RWL code is here and everyone can see how it works.

Reply:
The Finishers are removed.